src/server/pingserver-rs/src/worker.rs

```diff
@@ -73,74 +74,81 @@ impl Worker {
+                let pending = session.buffer().read_pending();
+                trace!("{} bytes pending in rx buffer for new session", pending);
+
+                // reserve vacant slab
+                let session_entry = self.sessions.vacant_entry();
+                let token = Token(session_entry.key());
 
-                    if let Some(session) = self.sessions.get_mut(token.0) {
-                        trace!(
-                            "{} bytes pending in rx buffer for session: {}",
-                            session.buffer().read_pending(),
-                            token.0
-                        );
-                        trace!(
-                            "{} bytes pending in tx buffer for session: {}",
-                            session.buffer().write_pending(),
-                            token.0
-                        )
+                // set client token to match slab
+                session.set_token(token);
+
+                // register tcp stream and insert into slab if successful
+                match session.register(&self.poll) {
+                    Ok(_) => {
+                        session_entry.insert(session);
+                        if pending > 0 {
+                            self.handle_data(token);
+                        }
                     }
-                } else {
-                    self.increment_count(&Stat::WorkerEventWake);
-                    // handle new connections
-                    while let Ok(mut s) = self
-                        .receiver
-                        .recv_timeout(std::time::Duration::from_millis(1))
-                    {
-                        let pending = s.buffer().read_pending();
-                        trace!("{} bytes pending in rx buffer for new session", pending);
-
-                        // reserve vacant slab
-                        let session = self.sessions.vacant_entry();
-                        let token = Token(session.key());
-
-                        // set client token to match slab
-                        s.set_token(token);
-
-                        // register tcp stream and insert into slab if successful
-                        match s.register(&self.poll) {
-                            Ok(_) => {
-                                session.insert(s);
-                                if pending > 0 {
-                                    self.handle_data(token);
-                                }
-                            }
-                            Err(_) => {
-                                error!("Error registering new socket");
-                            }
-                        };
+                    Err(_) => {
+                        error!("Error registering new socket");
+                    }
+                };
+            }
+
+            // poll queue to receive new messages
+            while let Ok(message) = self.message_receiver.try_recv() {
```

Comment:
handle messages before sessions for better request latency?

Reply:
Maybe?

The session may already have pending data to read - which we are handling immediately after registering the new session to the worker loop. I don't think I'd want to delay that processing until after the message queue is checked?

Definitely possible that data has arrived since that is not in the userspace buffer yet. But if the client hasn't already sent, it's possible it's just an idle connection in a pool? Whereas if we handle the messages before new sessions, we are definitely delaying the response for an eager client that sent a request already.

We should probably add some stats around this, would be an interesting thing to track.

Maybe a compromise is to set a bound on the number of messages we try to handle here? Since the only message is the shutdown message... it doesn't matter much currently. Any message will cause the thread to terminate.

Reply:
If a read event occurs on an existing socket, chances are that client is also “eager”, unless it’s an EOF/FIN which is less likely. Here the consideration is more about minimizing connection churn. By prioritizing (events on) fully established connections, we aim to reduce the chance of client timeouts on these connections and limit the churn to a smaller, more predictable subset of connections. 

Memcached has a fairness control mechanism to limit the number of processed requests per connection per event, we can reintroduce it if pipelining is more widely observed (eg if proxy is used commonly). But that works at a different level.

Reply:
So, in this design - we use the 'server' thread to drive the TCP listener - calling accept as necessary and driving any TLS handshake. By the time the connection is available on the queue for the worker to take ownership and register with its epoll, it's already fully established.

There are three cases related to this:
1 - worker receives a new fully established connection, the client has not sent data and will not for some time, the connection is registered into the 'worker' epoll loop and receives an event at some point in the future.
2 - worker receives a new fully established connection, this connection already has application data in the receive buffer (TLS early data, 0-rtt). This case is handled at https://github.com/twitter/pelikan/pull/309/files#diff-c1ae6312de0ac5c3c9fd29b44cd4d3304be7e5dbd66aceff207fea90d4045cf2R131 - we would not receive another readable event. 
3 - worker receives a new fully established connection, this connection has no application data in the receive buffer yet, but would not block if read were called immediately. This would be the result of a client sending application data while the 'worker' thread is handling other requests.

What I meant in my previous comment was that if we reorder message handling before session handling, we force that delay onto any early data request (assuming the worker is spending most of its time handling requests or waiting for poll to timeout).